cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/InMemoryServiceProgramRunner.java

```diff
@@ -0,0 +1,133 @@
+import co.cask.cdap.internal.app.runtime.ProgramOptionConstants;
+import co.cask.cdap.internal.app.runtime.ProgramRunnerFactory;
+import co.cask.cdap.internal.app.runtime.SimpleProgramOptions;
+import co.cask.cdap.proto.ProgramType;
+import com.google.common.base.Function;
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.google.common.collect.HashBasedTable;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.google.common.collect.Table;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.inject.Inject;
+import org.apache.twill.api.RunId;
+import org.apache.twill.internal.RunIds;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Map;
+
+/**
+ * For running {@link Service}. Only used in in-memory/standalone mode.
+ */
+public class InMemoryServiceProgramRunner extends AbstractInMemoryProgramRunner {
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryServiceProgramRunner.class);
+
+  private final String host;
+
+  @Inject
+  InMemoryServiceProgramRunner(CConfiguration cConf, ProgramRunnerFactory programRunnerFactory) {
+    super(programRunnerFactory);
+    this.host = cConf.get(Constants.AppFabric.SERVER_ADDRESS);
+  }
+
+  @Override
+  public ProgramController run(Program program, ProgramOptions options) {
+    // Extract and verify parameters
+    ApplicationSpecification appSpec = program.getApplicationSpecification();
+    Preconditions.checkNotNull(appSpec, "Missing application specification.");
+
+    ProgramType processorType = program.getType();
+    Preconditions.checkNotNull(processorType, "Missing processor type.");
+    Preconditions.checkArgument(processorType == ProgramType.SERVICE, "Only SERVICE process type is supported.");
+
+    ServiceSpecification serviceSpec = appSpec.getServices().get(program.getName());
+    Preconditions.checkNotNull(serviceSpec, "Missing ServiceSpecification for %s", program.getName());
+
+    //RunId for the service
+    RunId runId = RunIds.generate();
+    Table<String, Integer, ProgramController> components = startAllComponents(program, runId,
+                                                                              options.getUserArguments(), serviceSpec);
+    return new InMemoryProgramController(components, runId, program, serviceSpec, options.getUserArguments(),
+                                         ProgramRunnerFactory.Type.SERVICE_COMPONENT);
+  }
+
+  /**
+   * Creates and starts all components of the given Service program.
+   *
+   * @return A {@link Table} with component name in the row, instance id in the column and the {@link ProgramController}
+   *         of the component runner as the cell value.
+   */
+  // TODO: Move this method to AbstractInMemoryProgramRunner to reduce duplication of code.
```

Comment:
This comment is not longer applicable because of the refactor. This TODO can be removed.